step/v2/step.go

```diff
@@ -0,0 +1,264 @@
+// Step is a library for building user facing pipelines.
+//
+// Step is responsible for presenting each step in a job to the user, and halting the
+// pipeline on an error.
+package step
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/briandowns/spinner"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+//go:generate go run ./generate/calls.go
+
+type pipeline struct {
+	title     string
+	callstack []string
+	failed    error
+	spinner   *spinner.Spinner
+	labels    []string
+}
+
+type pipelineKey struct{}
+
+func withPipeline(ctx context.Context, p *pipeline) context.Context {
+	return context.WithValue(ctx, pipelineKey{}, p)
+}
+
+func getPipeline(ctx context.Context) *pipeline {
+	v := ctx.Value(pipelineKey{})
+	if v == nil {
+		return nil
+	}
+	return v.(*pipeline)
+}
+
+func Pipeline(name string, steps func(context.Context)) error {
```

Comment:
This may not be worth the effort to refactor, but I notice that pipelines are handling multiple concerns (eg. fancy display updates, recording a trace for replay). I think ideally these would be composble. So you could run a pipeline with tracing only or a pipeline with fancy display only or one with both, or you could swap out a different display feature that uses bubbletea, etc. 

Again, probably not worth spending a ton of time on it at the moment, but if it's not a difficult refactor, it would make the framework more flexible.